tempodb/tempodb.go

```diff
@@ -100,6 +101,9 @@ type readerWriter struct {
 	w backend.Writer
 	c backend.Compactor
 
+	uncachedReader backend.Reader
+	uncachedWriter backend.Writer
```

Comment:
It would be nice if we could shield off direct access to `r`, `w`, `uncachedReader` and `uncachedWriter`. So you are always forced to call `getReaderForBlock` and `getWriterForBlock`.
Unfortunately we can't shield private fields because everything is in the same package.

I think we should either add a comment to warn you or somehow combine these fields into a new struct that knows what to cache.

When I search for usage of `r` and `w` I still find some code using them directly (so without `getXxxForBlock`). I don't know if this is okay.

`r`:
- compactor.go:154 (`compact`)
- compactor.go:160 (`compact`)
- tempodb.go:208 (`CompleteBlock`)

`w`:
- compactor.go:251 (`appendBlock`)
- tempodb.go:208 (`CompleteBlock`)

Reply:
Some cases are unavoidable, such as compactor.go:154 to the block meta from the backend, so the fields will have to remain accessible.  compactor:160 and :251 are ok because they are iterating or flushing data, which does not involve the bloom filter.    tempodb.go:208 actually looks like obsolete code only used by a querier test.  The ingester used to call this when iterating the wal, but now it calls CompleteBlockWithBackend.

Reply:
Why is `compactor.go:154` unavoidable? Using `r` directly means we read using the cache, what's the difference with using the reader we get from `getReaderForBlock`?

Reply:
You're right.  I was thinking that there will be cases where the meta is not available, i.e. a call to get the meta, but in this case the compactor does already have it.